File: test/webui/update.test.js
'use strict';

// The update controls in the settings panel.
//
// Two buttons rather than one, because what they do is not the same kind
// of thing: checking asks GitHub what the latest release is, and
// installing replaces the program being used. Nothing happens on opening
// the panel — a check is an outbound request that says which version this
// machine runs, so it is asked for.

const test = require('node:test');
const assert = require('node:assert/strict');

const { load } = require('./harness');

const UP_TO_DATE = {
  current: '0.46.0', checked: true, latest: '0.46.0', available: false,
  can_install: true, detail: 'localcode 0.46.0 is the latest release',
};
const AVAILABLE = {
  current: '0.45.2', checked: true, latest: '0.46.0', tag: 'v0.46.0', available: true,
  can_install: true, asset: 'localcode-0.46.0-windows-amd64.msi', size: 23023104,
  page_url: 'https://github.com/o/r/releases/tag/v0.46.0',
};

async function settingsOpen(app) {
  app.el('settings-btn').click();
  await app.settle();
}

test('opening the settings panel asks GitHub nothing', async () => {
  const app = await load({ routes: { 'GET /api/update': AVAILABLE } });
  await settingsOpen(app);

  assert.equal(app.callsTo('GET', '/api/update').length, 0,
    'opening a panel should not tell GitHub which version this machine runs');
  assert.equal(app.el('update-note').textContent, '');
  assert.equal(app.el('update-install-btn').hidden, true);
});

test('a check with nothing new says so and offers no install', async () => {
  const app = await load({ routes: { 'GET /api/update': UP_TO_DATE } });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /latest release/);
  assert.equal(app.el('update-install-btn').hidden, true);
});

test('a newer release brings out the install button, naming what it will fetch', async () => {
  const app = await load({ routes: { 'GET /api/update': AVAILABLE } });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /0\.46\.0 is available/);
  assert.match(app.el('update-note').textContent, /localcode-0\.46\.0-windows-amd64\.msi/);
  assert.equal(app.el('update-install-btn').hidden, false);
  assert.match(app.el('update-install-btn').textContent, /0\.46\.0/);
});

// A daemon reached over the network says so, and the button stays away:
// installing there would replace the program on the server, at the
// request of a browser somewhere else.
test('a daemon that cannot install offers the release page instead', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...AVAILABLE, can_install: false,
        detail: 'install it on the machine running localcode, or from https://github.com/o/r/releases/tag/v0.46.0',
      },
    },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  assert.equal(app.el('update-install-btn').hidden, true);
  assert.match(app.el('update-note').textContent, /machine running localcode/);
});

// Replacing the running program is not undoable, so it is asked once
// plainly and a "no" does nothing at all.
test('installing asks first, and declining downloads nothing', async () => {
  const app = await load({
    routes: { 'GET /api/update': AVAILABLE, 'POST /api/update/install': { started: true } },
    confirm: false,
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.equal(app.callsTo('POST', '/api/update/install').length, 0);
});

test('installing reports what the daemon did', async () => {
  const app = await load({
    routes: {
      'GET /api/update': AVAILABLE,
      'POST /api/update/install': {
        version: '0.46.0', started: true, path: 'C:\\cache\\localcode-0.46.0-windows-amd64.msi',
        detail: 'the installer is running; localcode has to close for it to replace the files',
      },
    },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.equal(app.callsTo('POST', '/api/update/install').length, 1);
  assert.match(app.el('update-note').textContent, /installer is running/);
});

// GitHub being unreachable is a sentence beside the button, not a silence.
test('a check that fails says why', async () => {
  const app = await load({
    routes: { 'GET /api/update': { current: '0.45.2', checked: false, detail: 'dial tcp: no route to host' } },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /no route to host/);
  assert.equal(app.el('update-install-btn').hidden, true);
});

// The install that replaces localcode's own binary and brings it back.
//
// Before this the daemon replaced the binary, said "restart localcode to
// run the new version", and nothing did — so the version in the header
// stayed where it was and the update read as one that had not happened.
// The page's job is only to say what is going on: the connection is about
// to go away, and the browser reconnects to the new daemon on the same
// address by itself.
test('a restart is reported and no second install is offered', async () => {
  const app = await load({
    routes: {
      'GET /api/update': AVAILABLE,
      'POST /api/update/install': {
        version: '0.46.0', replaced: true, restarting: true, started: false,
        path: '/home/u/.cache/localcode/updates/localcode-0.46.0-linux-amd64.tar.gz',
        detail: 'installed over /home/u/.local/bin/localcode — restarting localcode now',
      },
    },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /restarting localcode now/);
  // Gone, not merely disabled: the daemon behind it is on its way out, and
  // a second click would post to a server that is no longer there.
  assert.equal(app.el('update-install-btn').hidden, true);
});

// A daemon reached from another machine replaces its own binary and stays
// running, because restarting it is not a browser's to order. The sentence
// is then the whole of what the user gets, so it has to arrive intact.
// An http mirror authenticates nothing, so the panel says so beside the
// address, on the offer itself rather than once somewhere else. The https
// mirror next to it names its source with no such mark.
test('a check against an http mirror marks the source unverified', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...AVAILABLE, source: 'http://mirror.internal/dl/', source_unverified: true,
      },
    },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /from http:\/\/mirror\.internal\/dl\//);
  assert.match(app.el('update-note').textContent, /unverified/);
  assert.match(app.el('update-note').textContent, /plain http/);
});

test('a check against an https mirror names the source with no unverified mark', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...AVAILABLE, source: 'https://mirror.internal/dl/',
      },
    },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /from https:\/\/mirror\.internal\/dl\//);
  assert.doesNotMatch(app.el('update-note').textContent, /unverified/);
});

// The install reply is read on its own, after the check has scrolled by,
// so it states the http source again rather than relying on the offer.
test('installing from an http mirror states the unverified source', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...AVAILABLE, source: 'http://mirror.internal/dl/', source_unverified: true,
      },
      'POST /api/update/install': {
        version: '0.46.0', verified: false, source_unverified: true,
        detail: 'installed over /home/u/.local/bin/localcode',
      },
    },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /unverified/);
  assert.match(app.el('update-note').textContent, /could not be verified/);
});

test('installing from an https mirror states no unverified source', async () => {
  const app = await load({
    routes: {
      'GET /api/update': AVAILABLE,
      'POST /api/update/install': {
        version: '0.46.0', verified: true,
        detail: 'installed over /home/u/.local/bin/localcode',
      },
    },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.doesNotMatch(app.el('update-note').textContent, /unverified/);
});

test('an install with no restart tells the user to restart', async () => {
  const app = await load({
    routes: {
      'GET /api/update': AVAILABLE,
      'POST /api/update/install': {
        version: '0.46.0', replaced: true, restarting: false, started: false,
        detail: 'installed over /home/u/.local/bin/localcode — restart localcode to run the new version',
      },
    },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /restart localcode to run the new version/);
});

// The install is asked once, in fixed words: the program in use is about
// to be replaced, and on Windows that means an elevation prompt and
// localcode closing. In the desktop window the window closes and comes
// back; anywhere else localcode has to be quit.
test('installing in the window asks in fixed words', async () => {
  const app = await load({
    routes: { 'GET /api/update': AVAILABLE },
    confirm: false,
    globals: { lcWindowCommand: () => {} },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.equal(app.confirmMessages.length, 1);
  assert.match(app.confirmMessages[0], /The window closes and the installer runs/);
  assert.match(app.confirmMessages[0], /opens again when the installer has finished/);
});

test('installing outside the window asks in fixed words', async () => {
  const app = await load({
    routes: { 'GET /api/update': AVAILABLE },
    confirm: false,
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.equal(app.confirmMessages.length, 1);
  assert.match(app.confirmMessages[0], /The installer starts when localcode exits/);
  assert.match(app.confirmMessages[0], /Quit localcode to run it/);
});

// In the desktop window the installer starts when the window closes,
// and LocalCode opens again when the installer has finished.
test('a window install says it starts on close and comes back', async () => {
  const app = await load({
    routes: {
      'GET /api/update': AVAILABLE,
      'POST /api/update/install': {
        version: '0.46.0', started: true,
        detail: 'The installer for localcode 0.46.0 starts when this window closes. LocalCode opens again when the installer has finished.',
      },
    },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /starts when this window closes/);
  assert.match(app.el('update-note').textContent, /opens again when the installer has finished/);
});

// In a terminal the installer starts when localcode exits, and the
// person has to quit it.
test('a terminal install says it starts on exit and must be quit', async () => {
  const app = await load({
    routes: {
      'GET /api/update': AVAILABLE,
      'POST /api/update/install': {
        version: '0.46.0', started: true,
        detail: 'The installer for localcode 0.46.0 starts when localcode exits. Quit localcode to run it.',
      },
    },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /starts when localcode exits/);
  assert.match(app.el('update-note').textContent, /Quit localcode to run it/);
});

// A recorded install that did not land is one line: the version, the
// exit code and what it means, and the log path.
test('a failed install is reported beside the offer', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...AVAILABLE,
        last_install: {
          version: '0.46.0', exit_code: 1602, status: 'cancelled', meaning: 'cancelled',
          log: 'C:\\Users\\u\\AppData\\Local\\localcode\\updates\\localcode-0.46.0-msi.log',
        },
      },
    },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  const note = app.el('update-note').textContent;
  assert.match(note, /0\.46\.0 did not install/);
  assert.match(note, /1602/);
  assert.match(note, /cancelled/);
  assert.match(note, /localcode-0\.46\.0-msi\.log/);
});

// A check with nothing new still reports the failed install.
test('a failed install is reported when already up to date', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...UP_TO_DATE,
        last_install: {
          version: '0.45.0', exit_code: 1603, status: 'failed', meaning: 'failed (exit code 1603)',
          log: '/home/u/.cache/localcode/updates/localcode-0.45.0-msi.log',
        },
      },
    },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  const note = app.el('update-note').textContent;
  assert.match(note, /latest release/);
  assert.match(note, /0\.45\.0 did not install/);
  assert.match(note, /1603/);
});

// A record with an installed status says the version was installed and
// that localcode has to be restarted to run it: a daemon that is still
// the old version reports the install it just staged. Only the
// not-installed statuses say "did not install".
test('an installed record says it was installed, not that it did not', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...AVAILABLE,
        last_install: {
          version: '0.46.0', exit_code: 0, status: 'installed', meaning: 'installed',
          log: 'C:\\Users\\u\\AppData\\Local\\localcode\\updates\\localcode-0.46.0-msi.log',
        },
      },
    },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  const note = app.el('update-note').textContent;
  assert.match(note, /0\.46\.0 was installed/);
  assert.match(note, /Restart localcode to run it/);
  assert.doesNotMatch(note, /did not install/);
});

test('a cancelled record still says it did not install', async () => {
  const app = await load({
    routes: {
      'GET /api/update': {
        ...AVAILABLE,
        last_install: {
          version: '0.46.0', exit_code: 1602, status: 'cancelled', meaning: 'cancelled',
          log: 'C:\\Users\\u\\AppData\\Local\\localcode\\updates\\localcode-0.46.0-msi.log',
        },
      },
    },
  });
  await settingsOpen(app);

  app.el('update-check-btn').click();
  await app.settle();

  const note = app.el('update-note').textContent;
  assert.match(note, /0\.46\.0 did not install/);
  assert.match(note, /cancelled/);
});

// In the desktop window the started installer gets its own sentence
// saying when this window closes, and the window closes itself.
test('a started window install says when it closes and closes it', async () => {
  const closed = [];
  const app = await load({
    routes: {
      'GET /api/update': AVAILABLE,
      'POST /api/update/install': {
        version: '0.46.0', started: true,
        detail: 'The installer for localcode 0.46.0 starts when this window closes. LocalCode opens again when the installer has finished.',
      },
    },
    globals: { lcWindowCommand: (cmd) => closed.push(cmd) },
  });
  await settingsOpen(app);
  app.el('update-check-btn').click();
  await app.settle();

  app.el('update-install-btn').click();
  await app.settle();

  assert.match(app.el('update-note').textContent, /This window closes in 3 seconds\./);
  await app.wait(3400);
  assert.deepEqual(closed, ['close']);
});

File: internal/daemon/update_test.go
package daemon

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"localcode/internal/events"
	"localcode/internal/update"
)

// A GitHub that answers with one release, so the check can be exercised
// without the internet and without a published version to match against.
func githubWith(t *testing.T, tag string) *httptest.Server {
	t.Helper()
	// Every asset a real release publishes, all nine of them, and the
	// Linux four are the reason this says so. They were missing, so a
	// daemon asked on Linux found nothing built for it: can_install came
	// back false and TestTheDesktopWindowIsOfferedTheInstall failed
	// there — on this platform only, silently, for as long as the gate
	// ran on one machine. The fixture, not the updater: `make dist`
	// writes all nine and check-dist.sh refuses a release missing any.
	v := strings.TrimPrefix(tag, "v")
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, `{"tag_name":%q,"html_url":"https://github.com/o/r/releases/tag/%s","body":"what changed","assets":[
			{"name":"localcode-%s-windows-amd64.msi","browser_download_url":"https://example/msi","size":10,"digest":"sha256:aa"},
			{"name":"localcode-%s-windows-arm64.zip","browser_download_url":"https://example/zip","size":10,"digest":"sha256:bb"},
			{"name":"LocalCode-%s-darwin-universal-app.tar.gz","browser_download_url":"https://example/app","size":10,"digest":"sha256:cc"},
			{"name":"localcode-%s-darwin-universal.tar.gz","browser_download_url":"https://example/tgz","size":10,"digest":"sha256:dd"},
			{"name":"localcode-%s-linux-amd64.deb","browser_download_url":"https://example/deb64","size":10,"digest":"sha256:ee"},
			{"name":"localcode-%s-linux-amd64.tar.gz","browser_download_url":"https://example/tgz64","size":10,"digest":"sha256:ff"},
			{"name":"localcode-%s-linux-arm64.deb","browser_download_url":"https://example/debarm","size":10,"digest":"sha256:11"},
			{"name":"localcode-%s-linux-arm64.tar.gz","browser_download_url":"https://example/tgzarm","size":10,"digest":"sha256:22"}
		]}`, tag, tag, v, v, v, v, v, v, v, v)
	}))
	t.Cleanup(srv.Close)
	return srv
}

func checkUpdate(t *testing.T, d *Daemon) map[string]any {
	t.Helper()
	rec := httptest.NewRecorder()
	d.Handler().ServeHTTP(rec, httptest.NewRequest(http.MethodGet, "/api/update", nil))
	if rec.Code != http.StatusOK {
		t.Fatalf("GET /api/update = %d: %s", rec.Code, rec.Body)
	}
	var body map[string]any
	if err := json.Unmarshal(rec.Body.Bytes(), &body); err != nil {
		t.Fatalf("decode: %v", err)
	}
	return body
}

func TestTheUpdateCheckReportsANewerRelease(t *testing.T) {
	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.45.2"
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	body := checkUpdate(t, d)
	if body["available"] != true {
		t.Errorf("available = %v, want true (0.46.0 is newer than 0.45.2)", body["available"])
	}
	if body["latest"] != "0.46.0" {
		t.Errorf("latest = %v", body["latest"])
	}
	if body["current"] != "0.45.2" {
		t.Errorf("current = %v", body["current"])
	}
}

func TestTheUpdateCheckSaysWhenThereIsNothingNew(t *testing.T) {
	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.46.0"
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	body := checkUpdate(t, d)
	if body["available"] != false {
		t.Errorf("available = %v, want false", body["available"])
	}
	if detail, _ := body["detail"].(string); !strings.Contains(detail, "latest release") {
		t.Errorf("detail = %q", detail)
	}
}

// A build from a working tree has no version to compare, and claiming it
// is up to date — or offering it a downgrade — would both be wrong.
func TestADevBuildIsNotOfferedARelease(t *testing.T) {
	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "dev"
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	body := checkUpdate(t, d)
	if body["available"] != false {
		t.Errorf("available = %v, want false for a dev build", body["available"])
	}
	if detail, _ := body["detail"].(string); !strings.Contains(detail, "not a release build") {
		t.Errorf("detail = %q, want it to say which build this is", detail)
	}
}

// The check is harmless anywhere; installing is not. A daemon reached over
// the network would be replacing the program on the *server* at the
// request of a browser somewhere else — the same rule as the folder
// picker, and the reason the panel has to be told which kind it is talking
// to rather than assuming.
func TestInstallingIsRefusedUnlessTheDaemonIsTheMachineInFrontOfYou(t *testing.T) {
	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.45.2"
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	if body := checkUpdate(t, d); body["can_install"] != false {
		t.Errorf("can_install = %v on a daemon that cannot install", body["can_install"])
	}

	rec := httptest.NewRecorder()
	d.Handler().ServeHTTP(rec, httptest.NewRequest(http.MethodPost, "/api/update/install", nil))
	if rec.Code != http.StatusForbidden {
		t.Fatalf("POST /api/update/install = %d, want 403: %s", rec.Code, rec.Body)
	}
	if !strings.Contains(rec.Body.String(), "releases") {
		t.Errorf("the refusal does not say where to get it: %s", rec.Body)
	}
}

func TestTheDesktopWindowIsOfferedTheInstall(t *testing.T) {
	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.45.2"
	d.AllowUpdateInstall = true
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	body := checkUpdate(t, d)
	if body["can_install"] != true {
		t.Errorf("can_install = %v, want true in the desktop window", body["can_install"])
	}
	if name, _ := body["asset"].(string); name == "" {
		t.Error("the check did not say which file it would install")
	}
}

// GitHub being unreachable is a sentence beside the button, not a failed
// request: the panel shows this to someone whose network is behind a
// proxy, and a status code tells them nothing.
func TestAFailedCheckAnswersWithTheReason(t *testing.T) {
	unreachable := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
	}))
	t.Cleanup(unreachable.Close)

	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.45.2"
	d.UpdateAPI = unreachable.URL

	body := checkUpdate(t, d)
	if body["checked"] != false {
		t.Errorf("checked = %v, want false", body["checked"])
	}
	if detail, _ := body["detail"].(string); detail == "" {
		t.Error("a check that failed said nothing about why")
	}
}

// And installing a release this build already has is refused rather than
// downloading and running an installer for the version already installed.
func TestInstallingWhatIsAlreadyInstalledIsRefused(t *testing.T) {
	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.46.0"
	d.AllowUpdateInstall = true
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	rec := httptest.NewRecorder()
	d.Handler().ServeHTTP(rec, httptest.NewRequest(http.MethodPost, "/api/update/install", nil))
	if rec.Code != http.StatusConflict {
		t.Fatalf("POST /api/update/install = %d, want 409: %s", rec.Code, rec.Body)
	}
}

// Replacing the binary and then leaving the old one running is the fault
// this pins: the install worked, the reply said so, and the version in
// the header never changed because nothing restarted. Whoever read that
// reasonably concluded the update had done nothing.
func TestAReplacedBinaryIsRestartedWhereItCanBe(t *testing.T) {
	replaced := update.Outcome{Replaced: true, Detail: "installed over /home/u/.local/bin/localcode"}

	detail, restarting := restartPlan(replaced, true)
	if !restarting {
		t.Error("a daemon that can restart itself did not, so the old binary keeps running")
	}
	if !strings.Contains(detail, "restarting") {
		t.Errorf("the reply does not say a restart is coming: %q", detail)
	}

	// The same install on a daemon reached from another machine. Restarting
	// it is not a browser's to order, so the sentence has to carry the
	// instruction instead of the process carrying it out.
	detail, restarting = restartPlan(replaced, false)
	if restarting {
		t.Error("a daemon with no restart hook reported a restart it cannot perform")
	}
	if !strings.Contains(detail, "restart localcode") {
		t.Errorf("the reply does not tell the user to restart: %q", detail)
	}
}

// An installer that is running replaces the files itself once localcode
// exits, and a .deb was only downloaded. Neither is this process being
// replaced, so neither gets a restart — nor a sentence about one.
func TestAnInstallThatReplacedNothingIsNotRestarted(t *testing.T) {
	for _, out := range []update.Outcome{
		{Started: true, Detail: "the installer is running"},
		{Detail: "downloaded to /tmp/x.deb — install it with: sudo apt install /tmp/x.deb"},
	} {
		detail, restarting := restartPlan(out, true)
		if restarting {
			t.Errorf("%q was treated as a replacement of the running binary", out.Detail)
		}
		if detail != out.Detail {
			t.Errorf("the answer was rewritten: %q became %q", out.Detail, detail)
		}
	}
}

// The check reports a recorded install that did not land: the version,
// the exit code and what it means, and the log path.
func TestTheCheckReportsAFailedInstall(t *testing.T) {
	dir := t.TempDir()
	log := dir + "/localcode-0.46.0-msi.log"
	if err := update.WriteMSIRecord(dir, update.MSIRecord{Version: "0.46.0", ExitCode: 1602, Log: log}); err != nil {
		t.Fatal(err)
	}
	old := msiRecordDir
	msiRecordDir = func() (string, error) { return dir, nil }
	t.Cleanup(func() { msiRecordDir = old })

	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.45.2"
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	body := checkUpdate(t, d)
	last, ok := body["last_install"].(map[string]any)
	if !ok {
		t.Fatalf("last_install = %v, want the recorded install", body["last_install"])
	}
	if last["version"] != "0.46.0" {
		t.Errorf("version = %v", last["version"])
	}
	if last["exit_code"] != float64(1602) {
		t.Errorf("exit_code = %v", last["exit_code"])
	}
	if meaning, _ := last["meaning"].(string); !strings.Contains(meaning, "cancelled") {
		t.Errorf("meaning = %q, want what 1602 means", meaning)
	}
	if last["log"] != log {
		t.Errorf("log = %v, want %q", last["log"], log)
	}
}

// The terminal install notice is a recovered error event: both clients
// draw that as a note that ends nothing. A plain error ends the TUI's
// turn while the daemon keeps running it, and the Web UI paints it as a
// failure for the same turn.
func TestTheTerminalInstallNoticeEndsNothing(t *testing.T) {
	const detail = "The installer for localcode 0.46.0 starts when localcode exits. Quit localcode to run it."
	ev := msiTerminalNotice(detail)
	if ev.Type != events.TypeError {
		t.Errorf("Type = %q, want the error event both clients already handle", ev.Type)
	}
	if errMsg, _ := ev.Data["error"].(string); errMsg != detail {
		t.Errorf("error = %q, want the install reply", errMsg)
	}
	if recovered, _ := ev.Data["recovered"].(bool); !recovered {
		t.Error("recovered is not true, so the TUI ends its turn while the daemon keeps running it")
	}
}

// An install that succeeded and is now the running version says nothing,
// and the record is cleared.
func TestTheCheckSaysNothingAboutTheRunningInstall(t *testing.T) {
	dir := t.TempDir()
	if err := update.WriteMSIRecord(dir, update.MSIRecord{Version: "0.46.0", ExitCode: 0, Log: dir + "/x.log"}); err != nil {
		t.Fatal(err)
	}
	old := msiRecordDir
	msiRecordDir = func() (string, error) { return dir, nil }
	t.Cleanup(func() { msiRecordDir = old })

	d := newTestDaemon(t, "http://127.0.0.1:1")
	d.Version = "0.46.0"
	d.UpdateAPI = githubWith(t, "v0.46.0").URL

	body := checkUpdate(t, d)
	if last, ok := body["last_install"]; ok && last != nil {
		t.Errorf("last_install = %v, want nothing for the running version", last)
	}
	if _, err := update.ReadMSIRecord(dir); err == nil {
		t.Error("the spent record was kept")
	}
}
